SentenceVectors.ecl

```diff
@@ -0,0 +1,651 @@
+    sentWords0 := JOIN(allWordsM, modD, LEFT.text = RIGHT.text,
+                        TRANSFORM(SentInfo, SELF.sentId := LEFT.sentId,
+                            SELF.vec := RIGHT.vec,
+                            SELF := LEFT),
+                        LOCAL);
+    // Redistribute by sentence id
+    sentWords := SORT(DISTRIBUTE(sentWords0, sentId), sentId, LOCAL);
+    SentInfo doRollup(sentInfo lr, SentInfo rr) := TRANSFORM
+      SELF.vec := lr.vec + rr.vec;
+      SELF.sentId := lr.sentId;
+      SELF.text := '';
+    END;
+    // Accumulate all of the word vectors into a single long vector
+    sentOut0 := ROLLUP(sentWords, doRollup(LEFT, RIGHT), sentId, LOCAL);
+    // Now reduce the concatenated word vectors to a single word vector, and restore the original
+    // sentence text.
+    sentD := DISTRIBUTE(sent, sentId);
+    sentOut := JOIN(sentOut0, sentD, LEFT.sentId = RIGHT.sentId, TRANSFORM(RECORDOF(LEFT),
+                              SELF.vec := Utils.calcSentVector(LEFT.vec, vecLen),
+                              SELF.text := RIGHT.text,
+                              SELF := RIGHT), LOCAL);
+    RETURN sentOut;
+  END;
+  /**
+    * Find the closest word(s) to each of a given set of test words.
+    * <p>Test words are specified as WordInfo records
+    * (i.e. words with their associated vectors).
+    * <p>Excludes the test word as a possible match.
+    * Returns N closest matches for each test word. Optionally includes
+    * NGrams (N > 1) as possible matches.
+    */
+  SHARED DATASET(Closest) closestWordVectors(DATASET(TextMod) mod, DATASET(WordInfo) words,
+                                            UNSIGNED1 N=1, BOOLEAN showNGrams=FALSE) := FUNCTION
+    // Similarity Record
+    simRec := RECORD
+      t_WordId id;
+      t_Word text;
+      t_Word otherWord;
+      REAL4 similarity;
+    END;
+    // Transform to calculate Cosine Similarity
+    simRec getSim(WordInfo word, TextMod modWord) := TRANSFORM
+      SELF.text := word.text;
+      SELF.similarity := Utils.cosineSim(word.vec, modWord.vec, vecLen);
+      SELF.id := word.wordId;
+      SELF.otherWord := modWord.text;
+    END;
+    // Only use the word portion of the model and optionally include NGrams.
+    modW := IF(showNGrams, mod(typ = t_ModRecType.word),
+                      mod(typ = t_ModRecType.word AND text[1] != '_'));
+    sim0 := JOIN(words, modW, LEFT.text != RIGHT.text, getSim(LEFT, RIGHT), ALL);
+    sim0D := DISTRIBUTE(sim0, id);
+    sim1 := SORT(sim0D, id, -similarity, LOCAL);
+    sim2 := DEDUP(sim1, id, KEEP N, LOCAL);
+    // Roll up the best matches to return on record for each test word, containing
+    // a set of best matches and the similarity for each.
+    sim3 := PROJECT(sim2, TRANSFORM(Closest, SELF.closest := [LEFT.otherWord],
+                                        SELF.similarity := [LEFT.similarity],
+                                        SELF := LEFT), LOCAL);
+    sim := ROLLUP(sim3, TRANSFORM(RECORDOF(LEFT),
```

Comment:
can probably be coded more efficiently, but only worth it if it is a bottleneck

Reply:
deferred